### DISCUSSION TITLE
AOP programming in C#

### DISCUSSION THREAD

**Discussion by HamedFathi** [dotnet/csharplang#605] [Language Ideas]
> In computing, aspect-oriented programming (AOP) is a programming paradigm that aims to increase modularity by allowing the separation of cross-cutting concerns. It does so by adding additional behavior to existing code (an advice) without modifying the code itself, instead separately specifying which code is modified via a "pointcut" specification, such as "log all function calls when the function's name begins with 'set'". This allows behaviors that are not central to the business logic (such as logging) to be added to a program without cluttering the code core to the functionality. AOP forms a basis for aspect-oriented software development.

I am just curios C# team specially @mattwar  has some ideas for supporting AOP in C#.
I rememberd @mattwar  had some proposal on AOP in somewhere (GITHUB) unfortunately I did not find it again.

I dont prefer using [Fody](https://github.com/Fody/Fody) or [Postsharp](https://www.postsharp.net) I hope C# support AOP natively. (We know Roslyn can not do that)

I like postsharp AOP model but I remembered Great Anders Hejlsberg was against it.
[http://www.postsharp.net/blog/post/Anders-Hejlsberg-Dead-Body](http://www.postsharp.net/blog/post/Anders-Hejlsberg-Dead-Body)

I want to know C# Team have some ideas or plans for supporting AOP or not.

what is your opinion ? Do you agree?

**Comment by HaloFour** [comment]
#107 should cover a lot of those use cases.

**Comment by svick** [comment]
@HamedFathi

If you quote [Wikipedia](https://en.wikipedia.org/wiki/Aspect-oriented_programming), I think you should make that clear, especially considering that a lot of people here know what AOP is.

> I dont prefer using Fody or Postsharp

Why not? Could you explain what your issues with those tools are?

**Comment by HamedFathi** [comment]
@svick

sorry you right!

about Fody I believe the idea behind fody is great but writing IL Code directly is not good and can be dangerous for future of the application I think Postsharp way to using attribute based AOP is better.

about postsharp all things is ok but this is commercial library and I think we can have free microsoft way for great AOP programming

**Comment by db-007** [comment]
@HamedFathi You are right, PostSharp is commercial.  I also wonder what is available for Xamarin.Forms?

**Comment by natalie-o-perret** [comment]
Nothing new under the sun?
